src/mm/device_alloc.rs

```diff
@@ -27,9 +28,33 @@ unsafe impl Allocator for DeviceAlloc {
 		assert!(layout.align() <= BasePageSize::SIZE as usize);
 		let size = layout.size().align_up(BasePageSize::SIZE as usize);
 
-		let virt_addr = ptr.as_ptr().expose_provenance();
-		let phys_addr = PhysAddr::from(virt_addr);
+		let phys_addr = PhysAddr::from(self.phys_addr(ptr.as_ptr()));
 
 		super::physicalmem::deallocate(phys_addr, size);
 	}
 }
+
+impl DeviceAlloc {
+	/// Returns a pointer corresponding to `phys_addr`.
+	pub fn ptr<T>(&self, phys_addr: usize) -> *mut T {
```

Comment:
Would `PhysAddr` instead of `usize` be an option here?

Reply:
Yes. I am a bit hesitant because I find the address types a bit unwieldy.

Many places (not all, though) want a physical address anyways so it makes sense to change both methods to work with `PhysAddr`. 👍